src/GafferArnoldUI/ArnoldLightVisualiser.cpp

```diff
@@ -0,0 +1,399 @@
+
+struct SurfaceTextureCacheGetterKey
+{
+	SurfaceTextureCacheGetterKey()
+		: shaderNetwork( nullptr ), resolution( Imath::V2i( 512 ) )
+	{
+	}
+
+	SurfaceTextureCacheGetterKey( const IECoreScene::ShaderNetwork::Parameter &out, const IECoreScene::ShaderNetwork *shaderNetwork, const Imath::V2i &resolution )
+		:	shaderNetwork( shaderNetwork ), resolution( resolution )
+	{
+		shaderNetwork->hash( hash );
+		hash.append( resolution.x );
+		hash.append( resolution.y );
+
+		output = out;
+		hash.append( output.shader );
+		hash.append( output.name );
+	}
+
+	operator const IECore::MurmurHash & () const
+	{
+		return hash;
+	}
+
+	IECoreScene::ShaderNetwork::Parameter output;
+	const IECoreScene::ShaderNetwork *shaderNetwork;
+	Imath::V2i resolution;
+	MurmurHash hash;
+};
+
+CompoundDataPtr surfaceTextureGetter( const SurfaceTextureCacheGetterKey &key, size_t &cost )
+{
+	cost = key.resolution.x * key.resolution.y * 3 * 4; // 3 x 32bit float channels;
+	return GafferOSL::ShadingEngineAlgo::shadeTexture( key.output, key.shaderNetwork, key.resolution );
+}
+
+typedef IECorePreview::LRUCache<IECore::MurmurHash, CompoundDataPtr, IECorePreview::LRUCachePolicy::Parallel, SurfaceTextureCacheGetterKey> SurfaceTextureCache;
+// Cache cost is in bytes
+SurfaceTextureCache g_surfaceTextureCache( surfaceTextureGetter, 1024 * 1024 * 64 );
+
+//////////////////////////////////////////////////////////////////////////
+// OSL query LRU cache
+//////////////////////////////////////////////////////////////////////////
+
+// TODO: Do we need to update this or get away with startup values?
+const char *g_oslSearchPaths = getenv( "OSL_SHADER_PATHS" );
+
+typedef std::shared_ptr<OSLQuery> OSLQueryPtr;
+
+struct OSLQueryCacheGetterKey
+{
+	OSLQueryCacheGetterKey()
+		:	shaderName()
+	{
+	}
+
+	OSLQueryCacheGetterKey( const std::string& name )
+		:	shaderName( name )
+	{
+		hash.append( name );
+	}
+
+	operator const IECore::MurmurHash & () const
+	{
+		return hash;
+	}
+
+	const std::string shaderName;
+	MurmurHash hash;
+};
+
+OSLQueryPtr oslQueryGetter( const OSLQueryCacheGetterKey &key, size_t &cost )
+{
+	cost = 1;
+
+	OSLQueryPtr result( new OSLQuery() );
+	if( result->open( key.shaderName, g_oslSearchPaths ? g_oslSearchPaths : "" ) )
+	{
+		return result;
+	}
+
+	return nullptr;
+}
+
+typedef IECorePreview::LRUCache<IECore::MurmurHash, OSLQueryPtr, IECorePreview::LRUCachePolicy::Parallel, OSLQueryCacheGetterKey> OSLQueryCache;
+OSLQueryCache g_oslQueryCache( oslQueryGetter, 128 );
+
+//////////////////////////////////////////////////////////////////////////
+// Network conversion helpers
+//////////////////////////////////////////////////////////////////////////
+
+// Re-connects any connections from oldName on oldShader to newName on newShader
+void remapOutputConnections( const InternedString &oldShader, const InternedString &oldName, const InternedString &newShader, const InternedString &newName, ShaderNetwork *network )
```

Comment:
I realised that now we only support 1:1 remapping (rather than splicing in multiple shaders), we dont need `oldShader`/`newShader`, so alternate improvement in https://github.com/GafferHQ/gaffer/pull/3491/commits/76c39c80e9826c74f8c2a1660b9a72870798336a